tests: add apparmor prompting integration tests

olivercalder opened the pull request:
Port the prompting-client integration tests to snapd and extend them to cover common use cases for AppArmor Prompting.

This work is tracked internally by https://warthogs.atlassian.net/browse/SNAPDENG-30450.

olivercalder commented:
Interesting that jammy tests failed but noble succeeded, and that the prompting-client integration tests #14387 pass on jammy. This suggests a problem with the test setup, rather than prompting support itself. The error would seem to suggest this as well, but I'm not sure:
```
Run the prompting client in scripted mode in the background
+ prompting-client.scripted --script=/home/test/integration-tests/tmp.shU6Z7onh1/script.json --grace-period=1 --var=BASE_PATH:/home/test/integration-tests/tmp.shU6Z7onh1
creating client
script path: /home/test/integration-tests/tmp.shU6Z7onh1/script.json
Error: Io(Os { code: 13, kind: PermissionDenied, message: "Permission denied" })
+ echo Attempt to write the file
Attempt to write the file
+ snap run --shell prompting-client.scripted -c echo it is written > /home/test/integration-tests/tmp.shU6Z7onh1/test.txt
/bin/bash: line 1: /home/test/integration-tests/tmp.shU6Z7onh1/test.txt: Permission denied
```

olivercalder commented:
Also want to add a test which queues up multiple requests, then the client replies to the final one, and it checks whether the previous requests are now handled by the reply. But this requires being able to have prompt entries in the script which do not have replies, and this is not yet supported, I believe.

olivercalder commented:
I think it's worth reviewing these as they are now, and either adding more cases here or opening follow-up PRs for more cases in the future.

Some remaining considerations for reviewers:
- Would it be preferable to move the `prompting-client.scripted` call to `task.yaml` instead of invoking it in each .sh file?
  - Sometimes the prompting-client call is expected to fail, so we would not want to universally check `success` in the result file... unless test cases which are expected to fail do manual checks of the result file and then mangle it to say `success` before returning
  - I think in every case I can think of we'll want to at least invoke it at the start, and any test case setup can occur as root without interfering with/generating any prompts
- Could test cases be moved to a loop within the test, rather than as distinct variants?
  - Currently, the `tests.session` is created and restored for every variant, which takes a while
  - If the tmpdir is created for each case, they should still stay isolated, that was the whole idea behind the top-level filter and using `mktemp`
- Should test cases be autodetected to include any `.sh` file in the directory?

maykathm commented on tests/main/apparmor-prompting-integration-tests/create_multiple_allow.sh:
```diff
@@ -0,0 +1,42 @@
+	snap run --shell prompting-client.scripted -c "echo $name is written > ${TEST_DIR}/${name}"
+done
+
+sleep 5 # give the client a chance to write its result and exit
```
Nitpick: If possible, I would try to get rid of all sleeps throughout this PR. Sleeps introduce unnecessary delay and mask the true wait condition. What about something like `timeout 5 bash -c 'while pgrep -f "prompting-client-scripted" > /dev/null; do true; done'`?

olivercalder replied:
I've replaced sleeps with timeouts like you suggested, and set the timeout duration as a variable in `task.yaml` which is passed into each test case script.

Thanks for the great suggestion about pgrep as well!

maykathm reviewed (approved):
Thanks! Overall, I think it looks great. To respond to your comments, I think it could be a little cleaner if, as you already mentioned, you moved the `prompting-client.scripted` call to the task level. Along that same vein, since each test has a custom action that it performs followed by a custom check for success, you could place "do action" and "check for success" logic for each test inside of functions that get called at the task level. That would slightly reduce code duplication since your wait conditions could also be moved to the task level. That said, I think the way you currently laid out the tests is just fine; I suppose I would more favor the change if you think many more tests will be added in the future.

olivercalder commented:
Problem seems to be the kernel locking up waiting for a reply?? I'm testing in a lxd VM. After sending the first write request, sending another one does not cause a kernel request, and trying to open another shell fails.

I'll record a video to show what I'm seeing...

olivercalder commented:
Here's the recording:

https://github.com/user-attachments/assets/83d306b6-5b6a-41a9-80b8-333fdab25f0a

Some annotations:
- At 5:40: Yes, the write did succeed, and `prompting-client.scripted` will exit successfully, it's just first waiting for the 10 second grace period to elapse to make sure that no unexpected prompts occur.
- At 6:30: We see responses sent back to the kernel for several previous requests. Those are the requests for test1.txt, test2.txt, test3.txt, plus two requests related to test4.txt (IDs 11 and 12): one for open/create (request ID 11) and one for actually writing the data to test4.txt (request ID 12). So snapd is doing what we expect it to be doing with respect to the outstanding prompts. But look: __*for all the other request IDs besides 11 and 12 (including the one for ID 6 which I incorrectly highlight), we see an error when responding: "error while responding to kernel: cannot perform IOCTL request APPARMOR_NOTIF_SEND: no such file or directory"*__. But the response for ID 12 occurs after other "no such file or directory" failures, so it's not like the listener socket doesn't exist: that message must refer to the file which we're trying to allow a write to, and that file has since timed out and failed to be created. The question is where/why the hang is occurring in the shell, which prevents us from kicking off subsequent writes. Since it's the shell itself hanging, I suspect it's something to do with the kernel rather than snapd locking up (which was an initial theory of mine). But if it's the kernel hanging, then I'm not sure why it wouldn't hang when we send a single request and the prompting-client.scripted sends back a response. So it seems to me that something about a subprocess of the shell (or in the case of the video, another shell spawned by tmux) being blocked on a syscall.
- At 7:30: To clarify, they're not receiving a response (at least from snapd) until well after they time out, so they're being denied either by the kernel directly or by the application (`snap run --shell firefox -c "..."`). We did see the explicit `apparmor="DENIED" ... name="test1.txt"` earlier at 2:35, so it seems to me it's the kernel blocking this.

I'm going to test another theory to see if I can bypass the shell locking up and get the test to behave. Another (shorter) video incoming...

olivercalder commented:
Okay, very interesting. Adding a sleep (to get the inner bash shells running first before anything hangs on the write) does *not* help, the same thing occurs. But we also saw an actual kernel trace this time:

Here's the recording, with some (overly-verbose, sorry) commentary about what's going on: https://bucket.calder.dev/integration-test-debugging/sleep-first.mp4

And here's the kernel trace:
```
Oct 09 22:10:45 integration kernel: INFO: task bash:19241 blocked for more than 122 seconds.
Oct 09 22:10:45 integration kernel:       Not tainted 6.8.0-45-generic #45-Ubuntu
Oct 09 22:10:45 integration kernel: "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
Oct 09 22:10:45 integration kernel: task:bash            state:D stack:0     pid:19241 tgid:19241 ppid:11904  flags:0x00000002
Oct 09 22:10:45 integration kernel: Call Trace:
Oct 09 22:10:45 integration kernel:  <TASK>
Oct 09 22:10:45 integration kernel:  schedule+0x33/0x110
Oct 09 22:10:45 integration kernel:  schedule_preempt_disabled+0x15/0x30
Oct 09 22:10:45 integration kernel:  rwsem_down_write_slowpath+0x27e/0x550
Oct 09 22:10:45 integration kernel:  ? step_into+0xfe/0x390
Oct 09 22:10:45 integration kernel:  down_write+0x5c/0x80
Oct 09 22:10:45 integration kernel:  open_last_lookups+0x137/0x400
Oct 09 22:10:45 integration kernel:  path_openat+0x99/0x2d0
Oct 09 22:10:45 integration kernel:  ? syscall_exit_to_user_mode+0x89/0x260
Oct 09 22:10:45 integration kernel:  do_sys_openat2+0xb3/0xe0
Oct 09 22:10:45 integration kernel:  __x64_sys_openat+0x55/0xa0
Oct 09 22:10:45 integration kernel:  x64_sys_call+0x1eb8/0x25c0
Oct 09 22:10:45 integration kernel:  do_syscall_64+0x7f/0x180
Oct 09 22:10:45 integration kernel:  ? irqentry_exit+0x43/0x50
Oct 09 22:10:45 integration kernel:  ? clear_bhb_loop+0x15/0x70
Oct 09 22:10:45 integration kernel:  ? clear_bhb_loop+0x15/0x70
Oct 09 22:10:45 integration kernel:  ? clear_bhb_loop+0x15/0x70
Oct 09 22:10:45 integration kernel: RIP: 0033:0x721bae97253b
Oct 09 22:10:45 integration kernel: RSP: 002b:00007ffec1a38490 EFLAGS: 00000246 ORIG_RAX: 0000000000000101
Oct 09 22:10:45 integration kernel: RAX: ffffffffffffffda RBX: 00005cfd2ac47260 RCX: 0000721bae97253b
Oct 09 22:10:45 integration kernel: RDX: 0000000000000241 RSI: 00005cfd2ac62c70 RDI: 00000000ffffff9c
Oct 09 22:10:45 integration kernel: RBP: 00005cfd2ac62c70 R08: 0000000000000000 R09: 0000000000000020
Oct 09 22:10:45 integration kernel: R10: 00000000000001b6 R11: 0000000000000246 R12: 0000000000000241
Oct 09 22:10:45 integration kernel: R13: 0000000000000000 R14: 00005cfd2ac62c70 R15: 0000721bae85d300
Oct 09 22:10:45 integration kernel:  </TASK>
```

Okay, I tried again, this time running each as `sudo -iu ubuntu snap run --shell "..."`, and it also didn't work, though I expected it would: https://bucket.calder.dev/integration-test-debugging/login-shell.mp4

Lastly, I tried with `sudo -iu ubuntu snap run --shell "sleep 10; ..."`, and that also didn't work. No nice kernel traces this time either. Nothing much new learned here, but here's the recording: https://bucket.calder.dev/integration-test-debugging/login-shell-sleep-first.mp4


So in conclusion, once one write is initiated (and not actioned by a prompt), it seems to lock up the shell and prevent other writes from being carried out. Interestingly, snapd and the prompting client are still able to respond to the kernel, but new requests aren't received by snapd from the kernel, and the shell seems to lock up. So I'm not sure what to make of this... I think I can add some more listener debug logging to make sure it's not snapd locking up, but since the shell locks up, I think it's something else.

olivercalder commented:
What I really need to do is try on oracular, as that has the fixes for the `.part` file download problem.

olivercalder commented:
I tried this on oracular, with basically the same result.

Here's the recording: https://bucket.calder.dev/integration-test-debugging/oracular.mp4

And here's the journalctl log, with `echo -n upcall > /sys/module/apparmor/parameters/debug` as well for extra logging: https://bucket.calder.dev/integration-test-debugging/oracular-log.txt

olivercalder commented:
Just noticed something else: the system locking up only occurs when prompting is blocked on a write, not on a read. And the things which are locked up also seem to be writes (though some writes do seem to be able to still occur, such as the client writing a response to the snapd API socket, or snapd writing a response to the kernel (using ioctl).

Here's the recording:

https://github.com/user-attachments/assets/663e15b5-64b0-4317-bd3a-d1868fa938c9

olivercalder commented:
Confirmed, I think: when I use bash instead of zsh, I can carry on doing things (e.g. `echo hello`) while the prompt for write is sitting open (blocking the write), but the moment I try to `touch foo`, that command hangs.

![image](https://github.com/user-attachments/assets/488a060d-e3fa-44d3-af31-485554a4f3fd)

olivercalder commented:
Oh, it's more interesting than that: it's that file creation *in the same directory* hangs, but other file creation succeeds!

In the file below, everything from `echo hello` onward happened after the write to `/home/oac/tag.png` was hanging. So I could `echo hello`, could read a file (in another directory), could write a file in another directory, but as soon as I tried to write a different file in the same directory, it hung. Very very interesting.

![image](https://github.com/user-attachments/assets/41be0dc2-d48d-4dcc-81f0-eb36a3a27258)

ZeyadYasser commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,100 @@
+    # unexpected prompt or an error from snapd.
+
+    echo "Run the prompting client in scripted mode in the background as the test user"
+    tests.session -u test exec prompting-client.scripted \
+        --script="${TEST_DIR}/script.json" \
+        --grace-period=1 \
+        --var="BASE_PATH:${TEST_DIR}" | tee "${TEST_DIR}/result" &
```
Does the prompting client keep running endlessly in the background?

How about using `systemd-run` to manage its life cycle instead?

olivercalder replied:
Ah I think you're right, if the test script exits early, the prompting client needs to be manually killed. I added this fix in one of the recent PRs.

Open to exploring systemd-run though...

ZeyadYasser replied:
Yes you can easily run as any user, the added benefit is that you don't need to handle pids yourself you let systemd handle that for you.

This is very similar to what you could use
https://github.com/canonical/snapd/blob/f9e69aa5441a3ae74d74804fdc9326c8d29ca5ce/tests/main/snap-remove-terminate/task.yaml#L56

ZeyadYasser reviewed (commented):
Thank you, looks really good! did a first pass.

Also, maybe the script files can go under a testdata directory so that they are not included in the go package? feel free to ignore this for now, it may make sense to do this for all tests anyway later.
https://pkg.go.dev/cmd/go#hdr-Package_lists_and_patterns
> Directory and file names that begin with "." or "_" are ignored by the go tool, as are directories named "testdata".

pedronis commented on tests/lib/tools/tests.session:
```diff
@@ -32,10 +32,6 @@ main() {
-			--)
-				shift
-				break
-				;;
```
why this change? seems a bit unrelated, also @sergiocazzolato should review

olivercalder replied:
The existing help text didn't actually work:

```
show_help() {
    echo "usage: tests.session exec [-u USER] [-p PID_FILE] [--] <CMD>"
    echo "       tests.session prepare | restore [-u USER | -u USER1,USER2,...]"
    echo "       tests.session kill-leaked"
    echo "       tests.session dump"
    echo "       tests.session has-system-systemd-and-dbus"
    echo "       tests.session has-session-systemd-and-dbus"
}
```

Here's the switch for the exec keyword:
```
            exec) 
                action="exec"
                shift
                # remaining arguments are the command to execute
                break
                ;;
```

So if you try to do `tests.session exec -u test -- echo foo`, it'll treat `-u`, `test`, `--` `echo` `foo` as the command to execute, which is wrong.

So I saw two options:
1. change the behavior of `exec` so it actually expects `--` and doesn't `break` immediately
2. remove the `--` from the `exec` help text

I checked every usage of `tests.session` in spread tests, and all of them put `-u USER -p PID_FILE` before the `exec`. So I did the latter with regards to `exec`. And `--` isn't used by any other command, so I thought it best to remove it.

sergiocazzolato replied:
Agree, the help needs an update, after the exec command it breaks, so what it expected is the cmd to execute.

sergiocazzolato replied:
other way to do this is removing the break in exec, this will force to use -- to delimit the command to execute.

pedronis commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,110 @@
+        --var="BASE_PATH:${TEST_DIR}" | tee "${TEST_DIR}/result" &
+
+    timeout 5 sh -xc 'while ! pgrep -f "prompting-client-scripted" > /dev/null; do sleep 0.1; done'
+    sleep 1 # wait for scripted client to start listening, otherwise it may miss the first prompt
```
this sleep seems fragile? can't we use something like lsof or something to see if the connection is there?

pedronis commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,110 @@
+    echo "Run the test script as the test user"
+    if ! tests.session -u test exec sh -x "${TEST_DIR}/${VARIANT}.sh" "$TEST_DIR" ; then
+        # kill the prompting-client-scripted process for this test run
+        pkill -f "prompting-client-scripted.*${TEST_DIR}"
```
shouldn't we always kill it in restore per above discussion?

olivercalder replied:
Yes, but would the scripted client running in the background prevent the `execute` phase from completing, and thus prevent `restore` from starting? In testing, I found that if the scripted client did not terminate during execute, the test would sit waiting until it timed out.

--------

Old answer when I hadn't caught the "in restore" in your question:

The prompting client scripted mode will terminate successfully after it has observed all expected prompts and the grace period elapses (so it's sure no unexpected prompts occurred after the last expected one). It will terminate unsuccessfully if it sees any unexpected prompt or if it receives an error after talking to the snapd API. I believe the only case when we need to kill it explicitly is if the shell script for the test case exited early for some reason.

pedronis replied:
probably worth expanding the comment on the pkill to mention that is still running because it might not have observed all the expected prompts?

pedronis commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,110 @@
+summary: Check that AppArmor Prompting works end-to-end
+
+details: |
```
this should have a a link to something that explains the scripting language of the prompting client?

olivercalder replied:
Sure, I'll ask Innes if there's documentation of the json format. Is that what you mean by scripting language? The only other script-related thing would be the `--var` argument, which performs bash-style variable substitution where `--var "FOO:BAR"` sets every instance of `$FOO` or `${FOO}` to `BAR` in the the script. Strange that there's no help text for that arg:

```
me@hostname:~$ prompting-client.scripted --help
Run a scripted client expecting a given sequence of prompts

Usage: prompting-client-scripted [OPTIONS] --script <FILE>

Options:
  -v, --verbose...              
  -s, --script <FILE>           The path to the input JSON file
      --var <VAR>               
  -g, --grace-period <SECONDS>  The number of seconds to wait following completion of the script to check for any unexpected additional prompts
  -h, --help                    Print help
```

olivercalder replied:
https://github.com/canonical/prompting-client/blob/main/docs/running-the-scripted-client.md

pedronis commented on tests/main/apparmor-prompting-integration-tests/read_single_allow.json:
```diff
@@ -0,0 +1,28 @@
+{
+  "version": 1,
+  "prompt-filter": {
+    "snap": "prompting-client",
+    "interface": "home",
+    "constraints": {
+      "path": "$BASE_PATH/.*"
+    }
+  },
```
what's the role of this outside of prompts?

olivercalder replied:
The overall purpose of the scripted client is to test that the correct prompts occur in the correct order. This top-level filter applies first, so only prompts matching this filter are considered when checking the order and contents of the prompts. For these tests, the top-level filter is a way of isolating the different test cases, so only prompts which have a path starting with the passed-in base path are checked against the script, and that base path is `mktemp`-unique to the test case.

pedronis reviewed (commented):
thank you for pushing this forward, did an high-level pass, and looked only at the first of the tests, there's quite a bit of timeout/sleep/retry based code, I'm worried about the behavior of that on slow testing machines. If we cannot do better, you should consider being at least a bit more pessimistic with the numbers, it might also make sense to code the most common wait/timeout times etc as env vars instead of having to change them one by one

sergiocazzolato commented on tests/lib/tools/tests.session:
```diff
@@ -1,7 +1,7 @@
 
 show_help() {
-	echo "usage: tests.session exec [-u USER] [-p PID_FILE] [--] <CMD>"
+	echo "usage: tests.session [-u USER] [-p PID_FILE] exec <CMD>"
```
I think the same applies for prepare/restore

sergiocazzolato commented on tests/lib/tools/tests.session:
```diff
@@ -1,7 +1,7 @@
 
 show_help() {
-	echo "usage: tests.session exec [-u USER] [-p PID_FILE] [--] <CMD>"
+	echo "usage: tests.session [-u USER] [-p PID_FILE] exec <CMD>"
```
we could have prepare and retore following the same logic
tests.session [-u USER | -u USER1,USER2,...]" prepare | restore

In fact, in most the tests it is used like this

sergiocazzolato replied:
We need just break after prepare/restore commands are found

sergiocazzolato commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,110 @@
+    what is expected, and that the client suffessfully allows the operations.
+
+systems:
+  - ubuntu-16.04-*
```
You can use:
  - ubuntu-16*
  - ubuntu-18*
  - ubuntu-2*
  - ubuntu-core-*

sergiocazzolato commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,110 @@
+prepare: |
+    tests.session prepare -u test
+    #shellcheck disable=SC2016
+    tests.session -u test exec sh -c 'mkdir -p "$HOME/integration-tests"' # $HOME must be that of test user
```
you can use /home/test instead of $HOME, it is done in many tests.

sergiocazzolato commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,100 @@
+    #shellcheck disable=SC2016
+    tests.session -u test exec sh -c 'rm -rf "$HOME/integration-tests"' # $HOME must be that of test user
+    tests.session restore -u test
+
```
the snap should be automatically removed during restore.

sergiocazzolato commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,110 @@
+    echo "outstanding prompts:"
+    snap debug api "/v2/interfaces/requests/prompts?user-id=$TEST_UID"
+    echo "rules:"
+    snap debug api "/v2/interfaces/requests/rules?user-id=$TEST_UID"
```
please add "|| true"

sergiocazzolato commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,110 @@
+        fi
+
+        # even if unsupported setting it to false should succeed
+        snap set system experimental.apparmor-prompting=false
```
I would add this in restore

sergiocazzolato commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,110 @@
+    snap set system experimental.apparmor-prompting=true
+
+    # Wait for snapd to begin restart
+    sleep 1
```
the retry is not enough?

olivercalder replied:
My concern was if the retry succeeded before the restart even began. But if we know the restart begins before the API request completes, we can rely on the retry. I'm sure if this is the case...

pedronis replied:
you could look at journalctl -u snapd  to know if a restart has happend I suppose, we might even already some helper to check things in logs?

olivercalder replied:
I used the MainPID check to ensure the PID has changed:
```diff
iff --git a/tests/main/apparmor-prompting-integration-tests/task.yaml b/tests/main/apparmor-prompting-integration-tests/task.yaml
index 64a7adaaef..f05e993fae 100644
--- a/tests/main/apparmor-prompting-integration-tests/task.yaml
+++ b/tests/main/apparmor-prompting-integration-tests/task.yaml
@@ -55,6 +55,10 @@ debug: |
     snap debug api "/v2/interfaces/requests/rules?user-id=$TEST_UID" || true
 
 execute: |
+    echo "Precondition check that snapd is active"
+    systemctl is-active snapd.service snapd.socket
+    SNAPD_PID="$(systemctl show --property MainPID snapd.service | cut -f2 -d=)"
+
     echo "Enable prompting via snap client where possible"
     # Prompting is unsupported everywhere but the Ubuntu non-core systems with
     # kernels which support apparmor prompting
@@ -72,10 +76,10 @@ execute: |
     echo "Enable AppArmor prompting experimental feature"
     snap set system experimental.apparmor-prompting=true
 
-    # Wait for snapd to begin restart
-    sleep 1
+    echo "Wait for snapd to begin restart"
+    retry --wait 1 -n 60 test "$SNAPD_PID" != "$(systemctl show --property MainPID snapd.service | cut -f2 -d=)"
 
-    echo "Wait until prompting is active"
+    echo "Wait until snapd is active"
     retry --wait 1 -n 60 systemctl is-active snapd
 
     echo "Check that shell script and scripted replies exist for $VARIANT"
```

sergiocazzolato reviewed (commented):
The change looks pretty good, some minor comments inline

olivercalder commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,100 @@
+    echo "Enable prompting via snap client where possible"
+    # Prompting is unsupported everywhere but the Ubuntu non-core systems with
+    # kernels which support apparmor prompting
+    if ! os.query is-ubuntu || os.query is-core || ! grep 'prompt' /sys/kernel/security/apparmor/features/policy/permstable32 ; then
```
Any Ubuntu kernel new enough to support the `prompt` permission should also support prompting. If a kernel doesn't because `features/policy/notify/` exists but doesn't contain a file called `user` with `file` in it, then something unexpected has happened in kernel land and we'd like to know.

The checks for `features/policy/notify/user` are because some future systems (non-Ubuntu, is my understanding) will use the same notify socket for the AppArmor complain redirect as is used for prompt messages, so we need a more robust check in snapd if that system is in place but prompting isn't supported. Any Ubuntu system prior to 24.10 does not have `features/policy/notify/` though, and prompting support is assumed by the presence of the `prompt` permission and the notify socket.

Basically, snapd in production is extra strict since it may run on systems with Ubuntu kernels, upstream kernels, custom kernels, etc., but in spread tests on Ubuntu, all systems with a new enough kernel (including jammy since it uses 6.5 in google cloud images iirc) should support prompting and we want to test them to catch if anything out of the ordinary happens.

(Sorry for the long-winded answer, writing this out helps me remind myself of the situation)

pedronis commented on tests/main/apparmor-prompting-integration-tests/read_single_allow.sh:
```diff
@@ -0,0 +1,31 @@
+
+TEST_DIR="$1"
+TIMEOUT="$2"
+if [ -z "$TIMEOUT" ] ; then
+	TIMEOUT=10
+fi
```
are you expecting people to invoke the scripts directly, so that's why this is here?

olivercalder replied:
If there's no second argument, then TIMEOUT would be `""` which would break the test, so I figured it would be safer to ensure TIMEOUT is set. And yes, I've found it useful to manually invoke these test case scripts when testing in a lxd VM, for example.

pedronis reviewed (approved):
still some comments but high-level +1 from me, please get at least a couple other relevant reviews

maykathm reviewed (approved):
Looks good!

ZeyadYasser commented on tests/main/apparmor-prompting-integration-tests/task.yaml:
```diff
@@ -0,0 +1,128 @@
+    echo "Enable prompting via snap client where possible"
+    # Prompting is unsupported everywhere but the Ubuntu non-core systems with
+    # kernels which support apparmor prompting
+    if ! os.query is-ubuntu || os.query is-core || NOMATCH 'prompt' < /sys/kernel/security/apparmor/features/policy/permstable32 ; then
```
`NOMATCH 'prompt' < /sys/kernel/security/apparmor/features/policy/permstable32` is not considered when `/sys/kernel/security/apparmor/features/policy/permstable32` doesn't exist.

Check: https://github.com/canonical/snapd/actions/runs/11386138513/job/31678432632?pr=14518#step:15:4157

olivercalder replied:
Good catch, thank you! I'll add this here and make sure the existing spread tests around prompting handle this correctly too.

olivercalder replied:
It seems `! grep 'prompt' /sys/kernel/security/apparmor/features/policy/permstable32` handles this properly even if the file doesn't exist, so I'm inclined to return to that instead. Or at least leave them be in the existing spread tests. I can do `! [ -f /sys/kerne/security/apparmor/features/policy/permstable32 ] || NOMATCH 'prompt' < /sys/kerne/security/apparmor/features/policy/permstable32`.

ZeyadYasser reviewed (commented):
Looks really good, Thank you!

I notice a bunch of spread failures, most notably:
- Unsupported systems escape the initial check https://github.com/canonical/snapd/actions/runs/11386138513/job/31678432632?pr=14518#step:15:4157
- Permission errors in 24.10 https://github.com/canonical/snapd/actions/runs/11386138513/job/31678434236?pr=14518#step:15:6449
- Many prepare failure, somehow timeout when installing the prompting-client

olivercalder commented:
I've moved the snap install to execute, after we ensure that snapd is active. And added a retry which should hopefully mitigate store flakeyness.

ZeyadYasser reviewed (approved):
Thank you!

olivercalder commented:
One solution to the store flakeyness around installing the prompting-client snap would be to cache the download in `SNAPD_WORK_DIR` and install it from the cached .snap file in the test prepare. However, there are rather a lot of dependencies for the snap, and I'd rather not hard-code what they are at this point in time, in case they change later.

This was on a fresh ubuntu 24.04 container in lxd, after updating the snapd deb to 2.65.3:
```
root@big-condor:~# snap download prompting-client
Fetching snap "prompting-client"
2024/10/21 14:32:50.885697 store_download.go:142: no host system xdelta3 available to use deltas
Fetching assertions for "prompting-client"
Install the snap with:
   snap ack prompting-client_51.assert
   snap install prompting-client_51.snap
root@big-condor:~# snap ack prompting-client_51.assert
root@big-condor:~# snap install prompting-client_51.snap
2024-10-21T14:33:23Z INFO Waiting for automatic snapd restart...
prompting-client 0+git.d60470b from Canonical✓ installed
root@big-condor:~# snap list
Name               Version          Rev    Tracking       Publisher   Notes
bare               1.0              5      latest/stable  canonical✓  base
core22             20240904         1621   latest/stable  canonical✓  base
gnome-42-2204      0+git.510a601    176    latest/stable  canonical✓  -
gtk-common-themes  0.1-81-g442e511  1535   latest/stable  canonical✓  -
prompting-client   0+git.d60470b    51     -              canonical✓  -
snapd              2.63             21759  latest/stable  canonical✓  snapd
```

An alternative approach would be to define a new test suite layer for AppArmor prompting integration tests, and move the snap install and `tests.session prepare` to the prepare of that suite, rather than the prepare of each test case. I would prefer this, as it removes assumptions about the prompting-client snap and simplifies the prepare/restore for each test case, but it may be difficult to introduce a new suite level here. @sergiocazzolato what do you think of this?

olivercalder commented:
Moved the prompting-client install back into prepare (rather than in execute), and rebased on master to pull in test fixes.

olivercalder commented:
Test failures:

Prepare failed:
- google-pro:ubuntu-fips-22.04-64:tests/fips/ --- unrelated

Execute failed:
- google:ubuntu-24.04-64:tests/main/upgrade-from-release --- unrelated, fixed on master
- google-core:ubuntu-core-18-64:tests/core/snapd-refresh-vs-services-reboots --- unrelated, Andrew has a pending fix
- google-core:ubuntu-core-18-64:tests/core/snapd-refresh-vs-services:start_w_2_49_2 --- unrelated, Andrew has a pending fix
- google-core:ubuntu-core-18-64:tests/core/snapd-refresh-vs-services:start_w_pr --- unrelated, Andrew has a pending fix
- google-core:ubuntu-core-18-64:tests/core/snapd-refresh-vs-services:start_w_stable --- unrelated, Andrew has a pending fix
- google-core:ubuntu-core-24-64:tests/core/mem-cgroup-disabled --- unrelated

Restore failed:
- google-distro-1:fedora-39-64:tests/main/component-from-store --- unrelated, fixed on master
- openstack:fedora-40-64:tests/main/component-from-store --- unrelated, fixed on master
- google-core:ubuntu-core-18-64:tests/core/snapd-refresh-vs-services-reboots --- unrelated, Andrew has pending fix
- google-core:ubuntu-core-24-64:tests/core/mem-cgroup-disabled --- unrelated

Plus all of the opensuse tests, none of which are related to this PR: https://github.com/canonical/snapd/actions/runs/11450399848/job/31907785826?pr=14518

----

This PR adds a new spread test and does not affect production code, so it's clear the other test failures are unrelated to the changes here. The changes to `tests.session` were verified to not affect other spread tests, as those tests already use the behavior we're now enforcing (and the other behavior was broken).
